src/collected_output.rs

```diff
@@ -7,7 +7,7 @@ use std::{
 
 pub(crate) struct Waiter {
     stdin: JoinHandle<io::Result<()>>,
-    stdout: JoinHandle<io::Result<Vec<u8>>>,
+    stdout: JoinHandle<io::Result<Option<Vec<u8>>>>,
```

Comment:
We have an invariant now: this will only be `None` when `stdout` is *not* captured and `Some(...)` when `stdout` is captured. We're introducing a new `Error::CradleBug` in case this invariant is ever unexpectedly broken.